### PR TITLE
Fix dependency download on resource update

### PR DESCRIPTION
**Pull request by sebglon**
### Description
Fix the dependency download on Terraform resource update: https://github.com/hashicorp/terraform-provider-helm/issues/576

### Acceptance tests
- [X] Have you added an acceptance test for the functionality being added?
- [X] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```
$ go clean -testcache && TESTARGS='-run=^TestAccResourceRelease_dependency$' make testacc  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test "./helm" -v -run=^TestAccResourceRelease_dependency$ -timeout 120m -parallel=4
=== RUN   TestAccResourceRelease_dependency
=== PAUSE TestAccResourceRelease_dependency
=== CONT  TestAccResourceRelease_dependency
    TestAccResourceRelease_dependency: provider_test.go:140: [DEBUG] Creating namespace terraform-acc-test-64w3w3hooo
Saving 1 charts
Downloading mariadb from repo https://kubernetes-charts.storage.googleapis.com/
Deleting outdated charts
E0908 08:37:33.228853  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:33.882876  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
Saving 1 charts
Downloading mariadb from repo https://kubernetes-charts.storage.googleapis.com/
Deleting outdated charts
E0908 08:37:47.458975  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:47.800902  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:48.118409  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:48.528200  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:52.012830  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
E0908 08:37:52.506606  388761 memcache.go:206] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
--- PASS: TestAccResourceRelease_dependency (24.39s)
PASS
ok      github.com/hashicorp/terraform-provider-helm/helm       24.497s
```

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-helm/blob/master/CHANGELOG.md):

```release-note
Fix dependency download on helm-release update.
```
### References

Related to issue: https://github.com/hashicorp/terraform-provider-helm/issues/576
Based on this PR: https://github.com/hashicorp/terraform-provider-helm/pull/578

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Review by rybus — APPROVED** [review]

**Comment by tillepille** [comment]
@rybus, any chance we get this merged anytime soon?
Just ran into this issue...

**Comment by rybus** [comment]
> @rybus, any chance we get this merged anytime soon?
> Just ran into this issue...

Wish I could but I ain't no admin ;)

**Comment by mcanevet** [comment]
@sebglon would it be possible to rebase this PR so that we have a chance to get it merged?

File: helm/resource_release.go [hunk 1]
```diff
@@ -459,30 +457,8 @@ func resourceReleaseCreate(ctx context.Context, d *schema.ResourceData, meta int
...
-					RepositoryCache:  m.Settings.RepositoryCache,
-				}
-				if err := man.Update(); err != nil {
-					return diag.FromErr(err)
-				}
-			} else {
-				return diag.FromErr(err)
-			}
-		}
+	if err = updateDependencyCharts(d, chart, path, m); err != nil {
```
Comment (sebglon) [inline helm/resource_release.go]:
```suggestion
	if err = updateDependencyCharts(d, c, path, m); err != nil {
```

File: helm/resource_release.go [hunk 2]
```diff
@@ -552,6 +528,37 @@ func resourceReleaseCreate(ctx context.Context, d *schema.ResourceData, meta int
...
+					return diag.FromErr(err)
+				}
+			} else {
+				return diag.FromErr(err)
+			}
+		}
+	}
+	return nil
+}
+
```
Comment (sebglon) [inline helm/resource_release.go]:
```suggestion
func updateDependencyCharts(d *schema.ResourceData, c *chart.Chart, path string, m *Meta) error {
	p := getter.All(m.Settings)

	updateDependency := d.Get("dependency_update").(bool)

	if req := c.Metadata.Dependencies; req != nil {
		// If CheckDependencies returns an error, we have unfulfilled dependencies.
		// As of Helm 2.4.0, this is treated as a stopping condition:
		// https://github.com/helm/helm/issues/2209
		if err := action.CheckDependencies(c, req); err != nil {
			if updateDependency {
				man := &downloader.Manager{
					Out:              os.Stdout,
					ChartPath:        path,
					Keyring:          d.Get("keyring").(string),
					SkipUpdate:       false,
					Getters:          p,
					RepositoryConfig: m.Settings.RepositoryConfig,
					RepositoryCache:  m.Settings.RepositoryCache,
				}
				if err := man.Update(); err != nil {
					return diag.FromErr(err)
				}
			} else {
				return diag.FromErr(err)
			}
		}
	}
	return nil
}

```

File: helm/resource_release_test.go [hunk 3]
```diff
@@ -894,7 +894,7 @@ func testAccHelmReleaseConfigPostrender(resource, ns, name, binaryPath string) s
 
 			set {
 				name = "serviceAccount.create"
-				value = false 
+				value = false
```
Comment (sebglon) [inline helm/resource_release_test.go]:
```suggestion
				value = false 
```

**Comment by sebglon** [comment]
@jrhouston can i have a feedback or can i help you to merge this PR?

**Comment by jrhouston** [comment]
I think it's good to go – just need a second 👀 from @dak1n1.

**Comment by dak1n1** [comment]
The code looks good, but I'm not sure if it's behaving as intended. Here's what I found for behavior:

* If the dependency (tgz file) doesn't exist, it's downloaded. This is good.
* If the dependency file doesn't exist at apply time, but its resources still exist on the cluster, those resources are deleted. This might not be as intended.
* On the third apply, since the dependency file exists now, the resource is re-created on the cluster.

Here's some output that demonstrates these behaviors.

All resources are created successfully, initially,  because all the tgz files are there.
```
$ ls umbrella-chart/charts/
dependency-bar-0.1.0.tgz  dependency-foo-0.1.0.tgz

$ terraform apply --auto-approve
Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

$ kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
test-dependency-bar-8645494dc8-r5mmx   1/1     Running   0          15s
test-dependency-foo-7657f9495b-zwbwk   1/1     Running   0          15s
test-umbrella-chart-6c57c4cf7d-npf6d   1/1     Running   0          15s
```

When I delete the charts directory, the associated resources are removed from the cluster. 

```
$ rm -rf umbrella-chart/charts

$ terraform apply --auto-approve

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

$ kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
test-umbrella-chart-6c57c4cf7d-npf6d   1/1     Running   0          3m3s
```

The files are successfully downloaded, but the resources are removed from the cluster during apply. Then, on a third apply, the resources are re-created from those tar files.

```
$ ls umbrella-chart/charts/
dependency-bar-0.1.0.tgz  dependency-foo-0.1.0.tgz

$ terraform apply --auto-approve
Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

$ kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
test-dependency-bar-8645494dc8-bk5r4   1/1     Running   0          22s
test-dependency-foo-7657f9495b-jpjlg   1/1     Running   0          22s
test-umbrella-chart-6c57c4cf7d-npf6d   1/1     Running   0          5m8s
```

**Comment by jrhouston** [comment]
This was a great catch @dak1n1. I did some digging and I could actually reproduce this problem on create too. It looks like what we were doing was:

1.  Loading the chart
2. Use it's metadata to check the dependencies
3. Run the equivalent of `helm dep update`
4. Run the install or upgrade action

_However_ because we had already loaded the chart in step 1, it didn't contain the updated dependencies so it was actually deleting them when we tried the `rm -rf charts/` reproducer. 

I've changed the code to load the chart, check the dependencies, then load the chart again if the dependencies have been updated. I've added a test that checks the underlying manifest to make sure this doesn't happen. 

Would appreciate a second 👀 just to confirm.

**Review by dak1n1 — APPROVED** [review]
Nice fix! This is working as expected now. I'm especially happy to see a regression test that successfully catches the error condition:

```
=== CONT  TestAccResourceRelease_dependency
    resource_release_test.go:977: Step 3/3 error: Check failed: 1 error occurred:
                * Check 1/4 error: expected 9 resources but got 3
```

(That's when I use the previous commit to test with). The current commit passes all tests. Good to go! 👍🏻

**Pull request merged**